posts/2024-09-05-impl-trait-capture-rules.md

```diff
@@ -0,0 +1,211 @@
+
+```rust
+let mut datums: Vec<Datum> = vec![Datum::default()];
+let iter = process_datums(&datums);
+datums.push(Datum::default()); // <-- Error!
+iter.next();
+```
+
+## Usability problems with this design
+
+The capture rules for `impl Trait` were decided early on based on a limited set of examples. Over time we have noticed a number of problems with them.
+
+### not the right default
+
+Surveys of major codebases (both the compiler and crates on crates.io) found that the vast majority of return-position impl trait values need to capture lifetimes, so the default behavior of not capturing is not helpful.
+
+### not sufficiently flexible
+
+The current rule is that return-position impl trait *always* captures type parameters and *sometimes* captures lifetime parameters (if they appear in the bounds). As noted above, this default is wrong because most functions actually DO want to capture lifetime parameters: that at least has a workaround (modulo some details we'll not below). But the default is also wrong because some functions do NOT want to capture type parameters, and there is no way to override that right now. The original intention was that [type alias impl trait](https://rust-lang.github.io/impl-trait-initiative/explainer/tait.html) would solve this use case, but that would be a very non-ergonomic solution (and stabilizing type alias impl trait is taking longer than anticipated due to other complications).
+
+### hard to explain
+
+Because the defaults are wrong, these errors are encountered by users fairly regularly, and yet they are also subtle and hard to explain (as evidenced by this post!). Adding the compiler hint to suggest `+ '_` helps, but it's not great that users have to follow a hint they don't fully understand.
+
+### incorrect suggestion
+
+Adding a `+ '_` argument to `impl Trait` may be confusing, but it's not terribly difficult. Unfortunately, it's often the wrong annotation, leading to unnecessary compiler errors -- and the *right* fix is either complex or sometimes not even possible. Consider an example like this:
+
+```rust
+fn process<T> {
+    context: &Context,
+    datums: Vec<T>,
+) -> impl Iterator<Item = ()> + '_ {
+    datums
+        .into_iter()
+        .map(|datum| context.process(datum))
+}
+```
+
+Here the `process` function applies `context.process` to each of the elements in `datums` (of type `T`). Because it captures `context`, it is declared as `+ '_`. This tells the compiler that the hidden type must outlive `'_`. Because this hidden type includes the `Vec<T>`, this in turn means that `T` must outlive `'_` -- but this is not true. So you get a compilation error ([try it on the playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=b742fbf9b083a6e837db0b170489f34a)). 
+
```

Comment:
```suggestion

The underlying problem with this suggestion is that it uses a convenient-but-sometimes-incorrect tool, an outlives bound, to work around the existing capture rules, which say a lifetime is captured only if it appears in the bounds of the `impl Trait`. Even when this technique fixes compilation errors, it creates a semantic confusion in the minds of users. A `+ 'a` bound  means "valid for at least as long as `'a`". But here, it is being used to mean that the hidden type _captures_ the lifetime, or in other words, it is "only valid for as long as `'a`". The fact that these have nearly opposite meanings makes it extremely difficult to build a mental model of what these bounds are actually _for_.

```

Reply:
I changed this text pretty significantly in the new revision. 

> Here the `process` function applies `context.process` to each of the elements in `datums` (of type `T`). Because the return value uses `context`, it is declared as `+ 'c`. Our real goal here is to allow the return type to use `'c`; writing `+ 'c` achieves that goal because `'c` not appears in the bound listing. However, writing `+ 'c` *also* means that the hidden type must outlive `'c` -- and in this case, that is not strictly necessary! The hidden type is going to be a wrapper around `std::vec::IntoIter<T>`, and for this type to outlive `'c`, we would need a where clause `T: 'c`. But that where-clause should only be required if we use a value of type `&'c T`, and we don't. The result is that this example doe snot compile ([try it on the playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=b742fbf9b083a6e837db0b170489f34a)), even though it should.

I'm going to see if I can merge some elements of what you wrote into that.